File: src/api/assessment_pool_fork.ts
// Browser contract for Assessment-owned immutable Question Pool fork operations.

import type { AssessmentEditNumber } from "../../generated/api/AssessmentEditNumber";
import type { AssessmentEntryId } from "../../generated/api/AssessmentEntryId";
import type { AssessmentQuestionPoolForkView } from "../../generated/api/AssessmentQuestionPoolForkView";
import type { AssessmentQuestionPoolSelectionCountReceipt } from "../../generated/api/AssessmentQuestionPoolSelectionCountReceipt";
import type { AssessmentId } from "../../generated/api/AssessmentId";
import type { AssessmentPointValue } from "../../generated/api/AssessmentPointValue";
import type { AssessmentEntryScoringRule } from "../../generated/api/AssessmentEntryScoringRule";
import type { CourseInstanceId } from "../../generated/api/CourseInstanceId";
import type { QuestionPoolId } from "../../generated/api/QuestionPoolId";
import type { QuestionPoolSelectedQuestionOrder } from "../../generated/api/QuestionPoolSelectedQuestionOrder";
import type { QuestionPoolEditNumber } from "../../generated/api/QuestionPoolEditNumber";
import type { PublishedQuestionRevisionTuple } from "../../generated/api/PublishedQuestionRevisionTuple";

/** Closed browser input for importing one reusable Pool into an Assessment-owned fork. */
export interface ImportAssessmentQuestionPoolForkInput {
  readonly sourceQuestionPoolId: QuestionPoolId;
  readonly authoredPosition: number;
  readonly selectionCount: number;
  readonly pointsPerItem: AssessmentPointValue;
  readonly selectedQuestionOrder: QuestionPoolSelectedQuestionOrder;
  readonly scoringRule: AssessmentEntryScoringRule;
}

/** Closed browser input for replacing current membership of an Assessment-owned Pool fork. */
export interface AppendAssessmentQuestionPoolForkMembersInput {
  readonly expectedQuestionPoolEditNumber: QuestionPoolEditNumber;
  readonly members: ReadonlyArray<PublishedQuestionRevisionTuple>;
  readonly interchangeabilityAttested: boolean;
}

/** Whole receipt after saving current membership of an Assessment-owned Pool fork. */
export interface AppendedAssessmentQuestionPoolForkMembers {
  readonly assessmentEntryId: AssessmentEntryId;
  readonly questionPoolEditNumber: QuestionPoolEditNumber;
  readonly assessmentEditNumber: AssessmentEditNumber;
}

/** Server-issued Assessment entry and fork Pool at Edit Number 1 after an atomic Pool import. */
export interface ImportedAssessmentQuestionPoolFork {
  readonly assessmentEntryId: AssessmentEntryId;
  readonly questionPoolId: QuestionPoolId;
  readonly questionPoolEditNumber: QuestionPoolEditNumber;
  readonly assessmentEditNumber: AssessmentEditNumber;
}

/** Narrow Assessment Pool-fork capability; it neither saves generic Assessment content nor issues Pool IDs. */
export interface AssessmentPoolForkClient {
  readonly getAssessmentQuestionPoolFork: (
    courseInstanceId: CourseInstanceId,
    assessmentId: AssessmentId,
    assessmentEntryId: AssessmentEntryId,
  ) => Promise<AssessmentQuestionPoolForkView>;
  readonly importAssessmentQuestionPoolFork: (
    courseInstanceId: CourseInstanceId,
    assessmentId: AssessmentId,
    input: ImportAssessmentQuestionPoolForkInput,
    expectedAssessmentEditNumber: AssessmentEditNumber,
  ) => Promise<ImportedAssessmentQuestionPoolFork>;
  readonly appendAssessmentQuestionPoolForkMembers: (
    courseInstanceId: CourseInstanceId,
    assessmentId: AssessmentId,
    assessmentEntryId: AssessmentEntryId,
    input: AppendAssessmentQuestionPoolForkMembersInput,
    expectedAssessmentEditNumber: AssessmentEditNumber,
  ) => Promise<AppendedAssessmentQuestionPoolForkMembers>;
  readonly updateAssessmentQuestionPoolSelectionCount: (
    courseInstanceId: CourseInstanceId,
    assessmentId: AssessmentId,
    assessmentEntryId: AssessmentEntryId,
    selectionCount: number,
    expectedAssessmentEditNumber: AssessmentEditNumber,
  ) => Promise<AssessmentQuestionPoolSelectionCountReceipt>;
}

File: crates/learning-data-access/src/assessment_pool_fork.rs
//! Trusted import of an immutable Question Pool fork for one Assessment Entry.

use async_trait::async_trait;
use question_model::{
    AssessmentEditNumber, AssessmentEntryId, AssessmentEntryScoringRule, AssessmentId,
    AssessmentPointValue, CourseInstanceId, QuestionId, QuestionPoolEditNumber,
    QuestionPoolSelectedQuestionOrder, QuestionRevisionTuple,
};

use crate::{SessionTokenHash, StoreError};

/// Server-issued Pool ID and source Pool ID for one Assessment-owned fork import.
///
/// This is deliberately not browser input: the server issues the child Pool ID
/// and the Store copies current source membership in one transaction.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ImportAssessmentPoolForkInput {
    pub course_instance_id: CourseInstanceId,
    pub assessment_id: AssessmentId,
    pub assessment_entry: AssessmentEntryId,
    pub expected_assessment_edit_number: AssessmentEditNumber,
    pub fork_question_pool_id: QuestionId,
    pub source_question_pool_id: QuestionId,
    pub authored_position: u32,
    pub selection_count: std::num::NonZeroU32,
    pub points_per_item: AssessmentPointValue,
    pub selected_question_order: QuestionPoolSelectedQuestionOrder,
    pub scoring_rule: AssessmentEntryScoringRule,
}

/// Immutable fork identity created and associated in one transaction.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ImportedAssessmentPoolFork {
    pub assessment_entry: AssessmentEntryId,
    pub question_pool_id: QuestionId,
    pub question_pool_edit_number: QuestionPoolEditNumber,
    pub assessment_edit_number: AssessmentEditNumber,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AppendAssessmentPoolForkMembersInput {
    /// Opaque route IDs; no browser supplies an internal Assessment ID.
    pub course_instance_id: CourseInstanceId,
    pub assessment_id: AssessmentId,
    pub assessment_entry: AssessmentEntryId,
    pub expected_assessment_edit_number: AssessmentEditNumber,
    pub expected_question_pool_edit_number: QuestionPoolEditNumber,
    pub members: Vec<QuestionRevisionTuple>,
    pub interchangeability_attested: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AppendedAssessmentPoolFork {
    pub assessment_entry: AssessmentEntryId,
    pub question_pool_edit_number: QuestionPoolEditNumber,
    pub assessment_edit_number: AssessmentEditNumber,
}

/// Session-authorized atomic Assessment Pool import boundary.
#[async_trait]
pub trait AssessmentPoolForkStore: Send + Sync {
    /// Creates a forked Pool at Edit Number 1 and associates it with one Assessment Entry.
    async fn import_assessment_question_pool_fork(
        &self,
        session_token_hash: SessionTokenHash,
        input: ImportAssessmentPoolForkInput,
    ) -> Result<ImportedAssessmentPoolFork, StoreError>;

    async fn append_assessment_question_pool_fork_members(
        &self,
        session_token_hash: SessionTokenHash,
        input: AppendAssessmentPoolForkMembersInput,
    ) -> Result<AppendedAssessmentPoolFork, StoreError>;
}
